File: K-thSmallestinLexicographicalOrder2.py
class Solution:
    def findKthNumber(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: int
        """
        def getCount(n, prefix):
            step = 1
            count = 0
            while prefix <= n:
                count += step
                prefix *= 10
                step *= 10
            if n < (prefix // 10 + step // 10 - 1):
                count -= step // 10 - (n - prefix // 10 + 1)
            return count

        def helper(n, k, prefix):
            for i in range(1 if prefix == 0 else 0, 10):
                num = getCount(n, prefix * 10 + i)
                if k <= num:
                    if k == 1:
                        return prefix * 10 + i
                    else:
                        return helper(n, k - 1, prefix * 10 + i)
                else:
                    k -= num
            return 0
        return helper(n, k, 0)


if __name__ == '__main__':
    n = int(input())
    k = int(input())
    solution = Solution()
    print(solution.findKthNumber(n, k))

File: K-thSmallestinLexicographicalOrder2.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    int findKthNumber(int n, int k) {
        int cur = 1;
        k--;
        while(k > 0) {
        	long long step = 0, first = cur, last = cur + 1;
        	while(first <= n) {
        		step += min((long long)n + 1, last) - first;
        		first *= 10;
        		last *= 10;
			}
			
			if(k >= step) {
				k -= step;
				cur++;
			}else {
				k--;
				cur *= 10;
			}
		}
		return cur;
    }
};

int main() {
	int n, k;
	cin>>n>>k;
	
	Solution *solution = new Solution();
	cout<<solution->findKthNumber(n, k);
	
	return 0;
}
